subsys/bluetooth/services/ots/ots_dir_list.c

```diff
@@ -0,0 +1,242 @@
+/*
+ * Copyright (c) 2021 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/types.h>
+#include <stddef.h>
+#include <string.h>
+#include <errno.h>
+#include <sys/dlist.h>
+#include <sys/byteorder.h>
+
+#include <bluetooth/services/ots.h>
+#include "ots_internal.h"
+#include "ots_obj_manager_internal.h"
+
+#include <logging/log.h>
+
+LOG_MODULE_DECLARE(bt_ots, CONFIG_BT_OTS_LOG_LEVEL);
+
+static struct bt_ots_dir_list dir_lists[CONFIG_BT_OTS_MAX_INST_CNT];
+
+static void dir_list_object_encode(struct bt_gatt_ots_object *obj,
+				     struct net_buf_simple *net_buf)
+{
+	uint8_t flags = 0;
+	uint8_t *record_length_p = net_buf->data;
+	uint16_t start_len = net_buf->len;
+
+	BT_OTS_DIR_LIST_SET_FLAG_PROPERTIES(flags);
+	BT_OTS_DIR_LIST_SET_FLAG_CUR_SIZE(flags);
+	if (obj->metadata.type.uuid.type == BT_UUID_TYPE_128) {
+		BT_OTS_DIR_LIST_SET_FLAG_TYPE_128(flags);
+	}
+
+	/* skip 16bits at the beginning of the record for the record's length */
+	net_buf_simple_add(net_buf, sizeof(uint16_t));
+
+	/* ID */
+	net_buf_simple_add_le48(net_buf, obj->id);
+
+	/* Name length */
+	net_buf_simple_add_u8(net_buf, strlen(obj->metadata.name));
+
+	/* Name */
+	net_buf_simple_add_mem(net_buf, obj->metadata.name,
+			       strlen(obj->metadata.name));
+
+	/* Flags */
+	net_buf_simple_add_u8(net_buf, flags);
+
+	/* encode Object type */
+	if (obj->metadata.type.uuid.type == BT_UUID_TYPE_16) {
+		net_buf_simple_add_le16(net_buf, obj->metadata.type.uuid_16.val);
+	} else {
+		net_buf_simple_add_mem(net_buf, obj->metadata.type.uuid_128.val,
+				       BT_UUID_SIZE_128);
+	}
+
+	/* encode Object Current size */
+	net_buf_simple_add_le32(net_buf, obj->metadata.size.cur);
+
+	/* Object properties */
+	net_buf_simple_add_le32(net_buf, obj->metadata.props);
+
+	/* Update the record length at the beginning */
+	sys_put_le16(net_buf->len - start_len, record_length_p);
```

Comment:
Instead of using a `start_len` variable, it'd be cleaner to store the pointer returned by the `net_buf_simple_add(net_buf, sizeof(uint16_t))` call on line 38 since that's the same as `net_buf->len - start_len`.

Reply:
Sorry, I mean store the value to record_length_p from the return value of `net_buf_simple_add()`

Reply:
Also, I don't think we have the habit of encoding the variable type in the variable name, so I'd drop the `_p` from the end

Reply:
Another option would be to use `net_buf_simple_reserve(net_buf, sizeof(uint16_t))` and then you can use `net_buf_simple_push_le16(net_buf, net_buf->len)` without the need for `start_len` or `record_length_p` at all. I think that would be the cleanest, assuming there's nothing else in the buffer (i.e. `net_buf->len` will be the amount of data that you've added)

Reply:
We need the `start_len` to calculate the length of the record once we are done inserting the values. 

We already store the pointer of when the object starts (`record_length_p`), but we don't have have the pointer to the tail (i.e. where the record ends), nor can I see any net_buf_simple API to get that. 

So I'm not sure how we can calculate the length of the record with pointers, instead of a start and end length.

Reply:
@Thalley see my comment that I posted almost simultaneously with yours. Does the buffer contain anything else in the beginning of this function?

Reply:
@Thalley I see this being called from a do-while loop, so I suppose the answer is "yes, the buffer may already contain other data"?

Reply:
@Thalley that said, I see you use the same "placeholder" pattern elsewhere. Please keep in mind the possibility of using `net_buf_simple_reserve()` + `net_buf_simple_push()` as an alternative pattern.

Reply:
It may. This function is called for each object, so the content of the buffer is basically an encoded array of records, each with a length field (as the first field per record). 

So we have 
`[ {length, field, field, field}, {length, field, field, field}, ...] `
where each record `{ ... }` is added per function call. 

Instead of `start_len` we could do 
```
uint8_t *start = net_buf_simple_add(net_buf, sizeof(uint16_t));
...
...
sys_put_le16(net_buf_simple_tail(net_buf) - start, start);
```

Would that be better?

Reply:
@Thalley yeah, I think that's a little better since it explicitly associates `start` and the call to `net_buf_simple_add()`, which initialising from `net_buf->data` doesn't do (at least as clearly)

Reply:
Agreed. Updated now :)